### PR TITLE
#2011 - Notification status and status_reason consistency

### PR DESCRIPTION
**Pull request by kalbfled**
# Description

These changes ensure that a notification's "status reason" is always the empty string when the status is "delivered."

The underlying ticket cited a problem without an explicit way to reproduce it, so I went looking for problems.  These changes cover all code paths as far as I can tell without a specific problem to fix.

issue #2011

## How Has This Been Tested?

This is not easy to test because I don't know how to force providers to return a particular status.  I'm relying heavily on unit tests.

- [x] new and modified unit tests pass
- [x] Send myself a text message, reply with STOP, and see that the status reason of the delivered message isn't overwritten with the "opted-out" blurb.  See the full documentation of my testing [here](https://github.com/department-of-veterans-affairs/notification-api/issues/2011#issuecomment-2397743760).

## Checklist

- [x] I have assigned myself to this PR
- [x] PR has an [appropriate title](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Engineering/team_agreement.md#naming-prs): #9999 - What the thing does
- [x] PR has a detailed description, including links to specific documentation
- [x] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Engineering/pull_request_labels.md#vanotify-labels) to the PR.
- [x] I did not remove any parts of the template, such as checkboxes even if they are not used
- [x] My code follows the [style guidelines](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Process/style_guide.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to any documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works. [Testing guidelines](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Process/testing_guide.md)
- [x] I have ensured the latest main is merged into my branch and all checks are green prior to review
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] The ticket was moved into the DEV test column when I began testing this change

### CODE REVIEW

File: app/celery/process_pinpoint_receipt_tasks.py [hunk 1]
```diff
@@ -133,11 +134,15 @@ def process_pinpoint_results(
                 notification.status_reason = 'The veteran responded with STOP.'
             elif record_status == 'OPTED_OUT':
                 notification.status_reason = 'The veteran is opted-out at the Pinpoint level.'
+        elif notification_status == NOTIFICATION_DELIVERED:
+            # Never include a status reason for a delivered notification.
+            notification.status_reason = ''
```
Comment (k-macmillan) [inline app/celery/process_pinpoint_receipt_tasks.py:139]: Same

File: app/celery/process_ses_receipts_tasks.py [hunk 2]
```diff
@@ -209,24 +210,9 @@ def process_ses_results(  # noqa: C901 (too complex 14 > 10)
...
-                NOTIFICATION_PERMANENT_FAILURE,
-            }
-        ):
-            # async from AWS means we may get a delivered status after a bounce, in rare cases
-            current_app.logger.warning(
-                'Notification: %s was marked as a bounce, cannot be updated to: %s',
-                notification.id,
-                notification_status,
-            )
-            return
```
Comment (k-macmillan) [inline app/celery/process_ses_receipts_tasks.py:229]: Why was this removed?

File: tests/app/celery/test_process_delivery_status_result_tasks.py [hunk 3]
```diff
@@ -313,3 +318,46 @@ def test_wt_delivery_status_callback_should_log_total_time(
...
+    )
+    callback_mock = mocker.patch('app.celery.process_delivery_status_result_tasks.check_and_queue_callback_task')
+
+    assert process_delivery_status(event=sample_delivery_status_result_message)
+    callback_mock.assert_called_once()
+
+    notify_db_session.session.refresh(notification)
+    assert notification.reference == 'SMhardcodedKWM'
+    assert notification.status == NOTIFICATION_DELIVERED
+    assert not notification.status_reason, 'This should be the empty string.'
```
Comment (k-macmillan) [inline tests/app/celery/test_process_delivery_status_result_tasks.py:363]: nit, not a required change.
You duplicate the reference and status reason a few times. Probably would have been best to create a data dictionary or use some setup variables at the top of the method.

File: tests/app/dao/notification_dao/test_notification_dao.py [hunk 4]
```diff
@@ -2034,39 +2039,42 @@ def test_update_notification_status_by_id_cannot_update_status_out_of_order_with
...
-
-    # attempt update without the condition current state
-    update_notification_status_by_id(notification_id=notification_list[0].id, status=next_status)
+    reference = str(uuid4())
+    initial_status_reason = '' if (current_status == NOTIFICATION_DELIVERED) else 'Because I said so!'
+    final_status_reason = initial_status_reason if (next_status == current_status) else 'just because'
 
-    # attempt update with the conditional current state
-    update_notification_status_by_id(
-        notification_id=notification_list[1].id, status=next_status, current_status=current_status
```
Comment (k-macmillan) [inline tests/app/dao/notification_dao/test_notification_dao.py:2062]: Not following why this was removed.

File: tests/app/dao/notification_dao/test_notification_dao.py [hunk 5]
```diff
@@ -2034,39 +2039,42 @@ def test_update_notification_status_by_id_cannot_update_status_out_of_order_with
...
+    else:
+        update_notification_status_by_id(
+            notification_id=notification.id, status=next_status, status_reason=final_status_reason
+        )
 
-    # get the notification object and make sure the values have changed
-    for ref in reference_tuple:
-        notification = dao_get_notification_by_reference(ref)
-        assert isinstance(notification, Notification)
-        assert notification.status == next_status
```
Comment (k-macmillan) [inline tests/app/dao/notification_dao/test_notification_dao.py:2069]: Same

**Review by k-macmillan — CHANGES_REQUESTED** [review]

File: app/celery/process_ses_receipts_tasks.py (hunk 2 above)
Comment (kalbfled) [reply app/celery/process_ses_receipts_tasks.py:229]: It's redundant to this previous "if" block.  The only different is a slightly different logged message.  Both blocks end with "return".

File: tests/app/dao/notification_dao/test_notification_dao.py (hunk 4 above)
Comment (kalbfled) [reply tests/app/dao/notification_dao/test_notification_dao.py:2062]: This is a consequence of my refactor to use the second parametrization.  The function is called below, on line 2063.  The side-by-side diff might make the reason for the change more obvious.

File: tests/app/dao/notification_dao/test_notification_dao.py (hunk 5 above)
Comment (kalbfled) [reply tests/app/dao/notification_dao/test_notification_dao.py:2069]: This is no longer appropriate because I added a new parametrization.

File: app/celery/process_ses_receipts_tasks.py (hunk 2 above)
Comment (kalbfled) [reply app/celery/process_ses_receipts_tasks.py:229]: We discussed this.  I think I get it.  `notification.status` refer to the existing status; not the incoming, updated status.

File: tests/app/dao/notification_dao/test_notification_dao.py [hunk 6]
```diff
@@ -2034,39 +2039,42 @@ def test_update_notification_status_by_id_cannot_update_status_out_of_order_with
...
-        # make sure we pass all the init setup
-        i = len(notification_list) - 1
-        assert isinstance(notification_list[i], Notification)
-        assert notification_list[i].status == current_status
-
-    # attempt update without the condition current state
-    update_notification_status_by_id(notification_id=notification_list[0].id, status=next_status)
+    reference = str(uuid4())
+    initial_status_reason = '' if (current_status == NOTIFICATION_DELIVERED) else 'Because I said so!'
+    final_status_reason = initial_status_reason if (next_status == current_status) else 'just because'
```
Comment (k-macmillan) [inline tests/app/dao/notification_dao/test_notification_dao.py:2050]: Please remove this. It always evaluates to false.

File: tests/app/dao/notification_dao/test_notification_dao.py (hunk 4 above)
Comment (k-macmillan) [reply tests/app/dao/notification_dao/test_notification_dao.py:2062]: I see how this was done with parametrization now, thank you!

File: app/celery/process_delivery_status_result_tasks.py [hunk 7]
```diff
@@ -194,20 +199,30 @@ def _get_notification_parameters(notification_platform_status: dict) -> Tuple[st
...
 ):
-    """Calculate pricing"""
+    """
+    Calculate pricing, and update the notification.
+    """
+
     current_app.logger.debug('Calculate pricing and update notification %s', notification.id)
+
+    # Delivered messages should not have an associated reason.
+    status_reason = '' if (notification_status == NOTIFICATION_DELIVERED) else notification.status_reason
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply app/celery/process_delivery_status_result_tasks.py:212]: Done.

File: app/celery/process_ses_receipts_tasks.py (hunk 2 above)
Comment (kalbfled) [reply app/celery/process_ses_receipts_tasks.py:229]: restored

File: tests/app/dao/notification_dao/test_notification_dao.py (hunk 6 above)
Comment (kalbfled) [reply tests/app/dao/notification_dao/test_notification_dao.py:2050]: done

File: tests/app/celery/test_process_delivery_status_result_tasks.py (hunk 3 above)
Comment (kalbfled) [reply tests/app/celery/test_process_delivery_status_result_tasks.py]: I didn't make any changes for this due to time constraints.

File: app/celery/process_delivery_status_result_tasks.py [hunk 8]
```diff
@@ -68,7 +70,7 @@ def process_delivery_status(
         price_in_millicents_usd,
     ) = _get_notification_parameters(notification_platform_status)
 
-    # retrieves the inbound message for this provider we are updating the status of the outbound message
+    # Retrieve the inbound message for this provider.  (We are updating the status of the outbound message.)
```
Comment (cris-oddball) [inline app/celery/process_delivery_status_result_tasks.py:73]: Nit: grammar. parenthetical statements go inside the sentence. it should be either
`# Retrieve the inbound message for this provider (we are updating the status of the outbound message).`

or 
`# Retrieve the inbound message for this provider. We are updating the status of the outbound message.`

**Review by cris-oddball — DISMISSED** [review]
approved with one grammatical nit.

File: app/celery/process_delivery_status_result_tasks.py (hunk 8 above)
Comment (kalbfled) [reply app/celery/process_delivery_status_result_tasks.py]: updated

**Review by cris-oddball — APPROVED** [review]

**Review by EvanParish — APPROVED** [review]
